plugins/elanfp/fu-elanfp-usb-device.c

```diff
@@ -0,0 +1,564 @@
+		return FALSE;
+
+	if (!g_usb_device_claim_interface(usb_device,
+					  ELANFP_USB_INTERFACE,
+					  G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					  error)) {
+		g_prefix_error(error, "failed to claim interface: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_elanfp_usb_device_close(FuDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+
+	if (!g_usb_device_release_interface(usb_device,
+					    ELANFP_USB_INTERFACE,
+					    G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					    error)) {
+		g_prefix_error(error, "failed to release interface: ");
+		return FALSE;
+	}
+
+	/* FuUsbDevice->close */
+	return FU_DEVICE_CLASS(fu_elanfp_usb_device_parent_class)->close(device, error);
+}
+
+static gboolean
+fu_elanfp_usb_device_do_xfer(FuElanfpUsbDevice *self,
+			     guint8 *outbuf,
+			     gsize outlen,
+			     guint8 *inbuf,
+			     gsize inlen,
+			     gboolean allow_less,
+			     gsize *rxed_count,
+			     GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	gsize actual = 0;
+
+	/* send data out */
+	if (outbuf != NULL && outlen > 0) {
+		if (!g_usb_device_bulk_transfer(usb_device,
+						ELAN_EP_CMD_OUT,
+						outbuf,
+						outlen,
+						&actual,
+						BULK_SEND_TIMEOUT_MS,
+						NULL,
+						error)) {
+			return FALSE;
+		}
+		if (actual != outlen) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_PARTIAL_INPUT,
+				    "only sent %" G_GSIZE_FORMAT "/%" G_GSIZE_FORMAT " bytes",
+				    actual,
+				    outlen);
+			return FALSE;
+		}
+	}
+
+	/* read reply back */
+	if (inbuf != NULL && inlen > 0) {
+		actual = 0;
+		if (!g_usb_device_bulk_transfer(usb_device,
+						ELAN_EP_IMG_IN,
+						inbuf,
+						inlen,
+						&actual,
+						BULK_RECV_TIMEOUT_MS,
+						NULL,
+						error)) {
+			return FALSE;
+		}
+		if (actual != inlen && !allow_less) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_PARTIAL_INPUT,
+				    "only received %" G_GSIZE_FORMAT "/%" G_GSIZE_FORMAT " bytes",
+				    actual,
+				    outlen);
+			return FALSE;
+		}
+	}
+
+	if (rxed_count != NULL)
+		*rxed_count = actual;
+
+	return TRUE;
+}
+
+static gboolean
+fu_elanfp_usb_device_setup(FuDevice *device, GError **error)
+{
+	FuElanfpUsbDevice *self = FU_ELANFP_USB_DEVICE(device);
+	gchar version[8] = {0};
+	guint8 usb_buf[2] = {0x40, 0x19};
+	gsize actual_len = 0;
+
+	fu_elanfp_usb_device_do_xfer(self,
+				     (guint8 *)&usb_buf,
+				     sizeof(usb_buf),
+				     usb_buf,
+				     sizeof(usb_buf),
+				     TRUE,
+				     &actual_len,
+				     error);
+
+	sprintf(version, "%02x%02x", usb_buf[0], usb_buf[1]);
+
+	g_debug("version %x %x", usb_buf[0], usb_buf[1]);
+
+	fu_device_set_version(device, version);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_elanfp_device_write_fw(FuElanfpUsbDevice *self, GBytes *fw, GError **error)
```

Comment:
I don;t think this needs to be split out as a new static function. Just include it in fu_elanfp_usb_device_write_firmware pls

Reply:
Fixed in commit 5af9c23594ac261d85fbf53fca02de08a5ef0150.